File: 2022/kotlin/src/main/kotlin/P01.kt
object P01 : Problem {
    override val problem = "p01"

    override val part1: (List<String>) -> Int = { input ->
        splitAtEmptyStrings(input, emptyList(), CalorieList(emptyList()))
            .map { it.toElf() }
            .maxOf { it.caloriesHeld }
    }
    override val part2: (List<String>) -> Int = { input ->
        splitAtEmptyStrings(input, emptyList(), CalorieList())
            .map { it.toElf() }
            .sortedBy { it.caloriesHeld }
            .reversed()
            .take(3)
            .sumOf { it.caloriesHeld }
    }

    private data class CalorieList(val calories: List<Int> = emptyList()) {
        fun toElf() = Elf(calories.sum())
    }

    private data class Elf(val caloriesHeld: Int)

    private fun splitAtEmptyStrings(
        strings: List<String>,
        completed: List<CalorieList> = emptyList(),
        curr: CalorieList = CalorieList(),
    ): List<CalorieList> =
        when {
            strings.isEmpty() -> completed + curr
            strings.first().isBlank() -> splitAtEmptyStrings(strings.tail(), completed + curr, CalorieList(emptyList()))
            else -> splitAtEmptyStrings(
                strings.tail(),
                completed,
                curr.copy(calories = curr.calories + strings.first().toInt())
            )
        }
}

File: 2022/haskell/p01.hs
import Data.List
import System.Environment
import Util

newtype CalorieList = CalorieList {calories :: [Int]} deriving (Show)

newtype Elf = Elf {caloriesCarried :: Int} deriving (Show, Eq, Ord)

calSum :: CalorieList -> Int
calSum calorieList = sum (calories calorieList)

main :: IO ()
main = getSolutions "P01" part1 part2 >>= \solution -> putStrLn solution

part1 :: String -> String
part1 input = show $ caloriesCarried $ maximum $ getElves input

parseCalorieLists :: [String] -> [CalorieList]
parseCalorieLists input = map mapStringsToCalorieList (splitBy "" input)

mapStringsToCalorieList :: [String] -> CalorieList
mapStringsToCalorieList input = CalorieList (map read input)

mapCaloriesListsToElves :: [CalorieList] -> [Elf]
mapCaloriesListsToElves = map (Elf . calSum)

getElves :: String -> [Elf]
getElves input = mapCaloriesListsToElves $ parseCalorieLists $ getLines input

part2 :: String -> String
part2 input = show $ sum $ toInts $ firstThree $ descendingSort $ getElves input

descendingSort = sortBy (flip compare)

firstThree :: [Elf] -> [Elf]
firstThree [] = []
firstThree [_] = []
firstThree [_, _] = []
firstThree (a : b : c : _) = [a, b, c]

toInts :: [Elf] -> [Int]
toInts = map caloriesCarried
